aea/helpers/dialogue/base.py

```diff
@@ -830,106 +849,88 @@ def update(self, message: Message) -> Optional[Dialogue]:
 
-        if not message.has_counterparty:
-            raise ValueError(
-                "The message counterparty field is not set {}".format(message)
-            )
-        if message.is_incoming and not message.has_sender:
-            raise ValueError("The message sender field is not set {}".format(message))
+        dialogue_reference = message.dialogue_reference
 
         is_invalid_label = (
-            dialogue_reference[0] == Dialogue.OPPONENT_STARTER_REFERENCE
-            and dialogue_reference[1] == Dialogue.OPPONENT_STARTER_REFERENCE
+            dialogue_reference[0] == Dialogue.UNASSIGNED_DIALOGUE_REFERENCE
+            and dialogue_reference[1] == Dialogue.UNASSIGNED_DIALOGUE_REFERENCE
         )
         is_new_dialogue = (
-            dialogue_reference[0] != Dialogue.OPPONENT_STARTER_REFERENCE
-            and dialogue_reference[1] == Dialogue.OPPONENT_STARTER_REFERENCE
+            dialogue_reference[0] != Dialogue.UNASSIGNED_DIALOGUE_REFERENCE
+            and dialogue_reference[1] == Dialogue.UNASSIGNED_DIALOGUE_REFERENCE
             and message.message_id == 1
-            and message.target == 0
-        )
-        is_incomplete_label_and_non_initial_msg = (
-            dialogue_reference[0] != Dialogue.OPPONENT_STARTER_REFERENCE
-            and dialogue_reference[1] == Dialogue.OPPONENT_STARTER_REFERENCE
-            and message.message_id > 1
         )
+
         if is_invalid_label:
             dialogue = None  # type: Optional[Dialogue]
-        elif is_new_dialogue and message.is_incoming:  # new dialogue by other
+        elif is_new_dialogue:  # initial message for new dialogue
             dialogue = self._create_opponent_initiated(
-                dialogue_opponent_addr=message.counterparty,
+                dialogue_opponent_addr=message.sender,
                 dialogue_reference=dialogue_reference,
                 role=self._role_from_first_message(message),
             )
-        elif is_new_dialogue and not message.is_incoming:  # new dialogue by self
-            dialogue = self._create_self_initiated(
-                dialogue_opponent_addr=message.counterparty,
-                dialogue_reference=dialogue_reference,
-                role=self._role_from_first_message(message),
-            )
-        elif (  # non-initial message with incomplete label
-            is_incomplete_label_and_non_initial_msg
-        ):
-            # we can allow a dialogue to have incomplete reference
-            # as multiple messages can be sent before one is received with complete reference
-            dialogue = self.get_dialogue(message)
-        else:
-            self._update_self_initiated_dialogue_label_on_message_with_complete_reference(
-                message
-            )
+        else:  # non-initial message for existing dialogue
+            self._complete_dialogue_reference(message)
             dialogue = self.get_dialogue(message)
```

Comment:
keep incomplete label logic!

Reply:
Discussed. It is not reached here given update is now only for incoming messages. 

The scenario where multiple messages are sent before one is received now happens with `Dialogue.reply`, for which there is an associated test now.

Reply:
Actually, on consideration, this api is still available so it needs to support the case where it is used directly without reply. We don't enforce usage of `reply`, we just make it available after all.

Reply:
This would only add further messiness to this already bloated module, where we support multiple ways of achieving the same thing, and leads to repetition of code, obscure errors, ...

Reply:
If we are making breaking changes, as its result developers are to update their code, then we should create the clean state we are after. Otherwise, when is a good time to make this separation of responsibilities between the three methods (reply, create, update) if not when we are doing a major restructuring?